File: AirSim/project/airsim_2dnav/scripts/cpp/goals.cpp
#include "ros/ros.h"
#include "geometry_msgs/Vector3.h"
#include "nav_msgs/Odometry.h"
#include <vector>
using std::vector;

class Goal
{
private:
  int goal_cnt;
  int config;
  vector<int> goal_types;
  vector<vector<double>> pose_seq;
  ros::Publisher cont_pub;
  ros::Subscriber sub_odom;
public:
  Goal(ros::NodeHandle *n) {
    goal_cnt = 0;
    config = 10;
    cont_pub = n->advertise<geometry_msgs::Vector3>("/goals", 1);
    sub_odom = n->subscribe("/odom", 5, &Goal::odom, this);
  }

  void callback();
  void odom(const nav_msgs::Odometry::ConstPtr& msg);
};

void Goal::callback() {
  ROS_INFO("Loading goals for configuration: %d", config);
  while (cont_pub.getNumSubscribers() == 0){
    if (ros::ok()) {
      ros::Duration(0.05).sleep();
    }
    else {
      return;
    }
  }
  // ros::Duration(1).sleep();

  switch (config) {
    case 1:
        // # Config 1 - Default (~1540 m, ~167s)
        pose_seq = {{-80, 0}, {-135, 0.5}, {-205, 0}, {-212, -15}, {-212, -30}, {-212, -74}, {-212,-122},{-197,-128},{-179,-128},{-135.5,-128},{-90,-128},{-84,-143},{-84,-153},{-84,-175},{-83,-195},{-79,-203},{-78, -207},{-84, -208}, {-85, -205}, {-85,-195},{-84,-128},{-84,-68},{-84,-58},{-99,-48},{-115,-48},{-135,-48},{-203, -48},{-212, -63},{-212, -78},{-212,-198},{-212, -241},{-212, -246},{-197, -256},{-182, -256},{-135.5, -256},{-80,-256},{-20, -256},{34.5, -256},{44, -241},{44, -220},{44, -198},{44, -138},{29, -128},{1,-128},{7.5, -133},{15, -128},{37, -128},{44, -110}, {44,-95}, {44.5, -65.5}};
        goal_types = {0,4,1,2,3,4,1,2,3,4,1,2,3,4,1,5,5,5,5,3,0,1,1,2,3,4,1,2,3,4,1,1,2,3,4,0,4,1,2,3,4,1,2,7,6,5,1,2,3,7};
        break;
    case 2:
        // # Config 2 - Efficiency (< Distance, < Time), (~1520 m, ~162s)
        pose_seq = {{-80, 0}, {-135, 0.5}, {-205, 0}, {-212, -15}, {-212, -30}, {-212, -74}, {-212,-120},{-197,-128},{-179,-128},{-135.5,-128},{-94,-128},{-84,-143},{-84,-153},{-84,-198},{-84,-246},{-99,-256}, {-114, -256}, {-135.5, -256}, {-202, -256}, {-212, -241}, {-212, -226},{-212,-198},{-212, -57},{-197, -48}, {-183,-48}, {-135,-48}, {-94,-48}, {-84,-63}, {-84,-78}, {-84,-120}, {-69,-128}, {-54,-128}, {0,-128},{34, -128},{44, -110}, {44,-95}, {44, -90}, {43,-79},{39,-71},{38, -67},{44, -66}, {45, -69}, {45,-77},{44, -198},{44, -246},{29, -256},{14,-256},{0, -256}};
        goal_types = {0,4,1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,1,2,3,4,1,2,3,4,1,5,5,5,5,3,4,1,2,3,7};
        break;
    case 3:
        // # Config 3 - Fastest (> Distance, <<< Time), (~1555 m, ~144s)
        pose_seq = {{-80, 0}, {-135, 0}, {-205, 0}, {-212, -15}, {-212, -30}, {-212, -74}, {-212,-120},{-197,-128},{-179,-128},{-135.5,-128}, {-70, -128}, {0,-128},{34, -128},{44, -110}, {44,-95}, {44, -79}, {43,-75}, {39, -67},{44, -66}, {45, -69}, {45,-77},{44, -198},{44, -246},{29, -256},{14,-256},{0, -256},{-70,-256}, {-135.5, -256}, {-202, -256}, {-212, -241}, {-212, -226},{-212,-198},{-212, -57},{-197, -48}, {-183,-48}, {-135,-48}, {-94,-48}, {-84,-63}, {-84,-78}, {-84,-198}};
        goal_types = {0,4,1,2,3,4,1,2,3,4,0,4,1,2,3,4,1,5,5,5,3,4,1,2,3,4,0,4,1,2,3,4,1,2,3,4,1,2,3,7};
        break;
    case 4:
        // # Config 4 - Shortest (<<< Distance, >> Time), (~1440 m, ~180s)
        pose_seq = {{-80, 0}, {-135, 0.5}, {-203, 0}, {-212, -15}, {-212, -30},{-212, -38},{-197, -48}, {-183,-48}, {-163,-48}, {-141,-49},{-133,-51},{-130, -48},{-144, -48}, {-153, -48},{-202, -48},{-212, -63},{-212, -78},{-212, -74}, {-212,-120},{-197,-128},{-179,-128}, {-163,-128}, {-141,-129},{-133,-131},{-130, -128},{-143, -128},{-153, -128},{-202,-128}, {-212,-143}, {-212, -158}, {-212,-198},{-212, -246},{-197, -256},{-182, -256},{-135.5, -256},{-94,-256},{-84,-241},{-84,-226},{-84.5,-198},{-84,-140},{-72,-128}, {-55,-128}, {0,-128},{34, -128},{44, -110}, {44,-95}, {44, -90}, {43,-78},{39,-69},{38, -67},{44, -65}, {45, -67}, {44,-85},{44, -198},{44, -246},{29, -256},{14,-256},{0, -256}};
        goal_types = {0,4,1,2,3,1,2,3,4,1,5,5,5,3,1,2,3,4,1,2,3,4,1,5,5,5,3,1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,4,1,5,5,5,5,3,4,1,2,3,7};
        break;
    case 5:
        // # Config 5 - Shortest (<<<< Distance, >> Time), (~1398 m, ~215s) (reverse, forward, reverse)
        pose_seq = {{-80, 0}, {-135, 0.5}, {-203, 0}, {-212, -15}, {-212, -30},{-212, -38},{-197, -48}, {-183,-48}, {-135,-48},{-202, -48},{-212, -63},{-212, -78},{-212, -74}, {-212,-120},{-197,-128},{-179,-128}, {-135,-128},{-202,-128}, {-212,-143}, {-212, -158}, {-212,-198},{-212, -246},{-197, -256},{-182, -256},{-135.5, -256},{-94,-256},{-84,-241},{-84,-226},{-84.5,-198},{-84,-140},{-74,-128}, {-59,-128}, {0,-128},{34, -128},{44, -110}, {44,-95}, {44, -65},{44, -198},{44, -246},{29, -256},{14,-256},{0, -256},{0, -256}};
        goal_types = {0,4,1,2,3,1,2,3,4,8,9,10,6,8,9,10,6,1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,4,1,2,3,4,6,8,9,10,6,7};
        break;
    case 6:
        // # Config 6 - Shortest (<<<< Distance, >>> Time), (~1400 m, ~200s) (reverse, forward, reverse) x2
        pose_seq = {{-80, 0}, {-135, 0.5}, {-203, 0}, {-212, -15}, {-212, -30},{-212, -38},{-197, -48}, {-183,-48}, {-135,-48},{-199, -48},{-212, -65},{-212, -78},{-212, -74}, {-212,-116},{-197,-128},{-179,-128}, {-135,-128},{-202,-128}, {-212,-143}, {-212, -158}, {-212,-198},{-212, -246},{-197, -256},{-182, -256},{-135.5, -256},{-94,-256},{-84,-241},{-84,-226},{-84.5,-198},{-84,-244},{-69,-256},{-54, -256}, {0, -256}, {31, -256}, {44, -241}, {44, -226}, {44, -198}, {44, -138},{29, -128},{1,-128},{34, -128},{44, -110}, {44,-95}, {44.5, -65}, {44.5, -65}};
        goal_types = {0,4,1,2,3,1,2,3,4,6,8,6,6,6,8,6,6,1,2,3,4,1,2,3,4,1,2,3,4,6,8,6,6,6,8,6,6,6,8,6,1,2,3,1,7};
        break;
    case 7:
        // # Config 7 - Start backwards
        pose_seq = {{0,0},{35, 0},{44, -15},{44,-30},{44,-65},{44,-95},{44,-110},{34,-127},{2,-129},{34,-129},{44,-143},{44,-158},{44,-198},{44, -246},{29, -256},{14,-256},{0, -256},{-74,-256},{-84,-241},{-84,-226},{-84.5,-200},{-84,-226},{-84,-241},{-94,-256},{-135.5,-256},{-182,-256},{-197,-256},{-210,-246},{-212,-195},{-212,-158},{-212,-143},{-202,-130},{-137,-128},{-202,-128},{-212,-113},{-212,-98},{-212,-74},{-212,-78},{-212,-63},{-200,-49},{-137,-48},{-183,-48},{-197,-48},{-210,-37},{-212,-30},{-212,-15},{-205,-1},{-180,0.0},{-135,1},{-135,1}};
        goal_types = {0,8,9,10,6,6,9,10,6,1,2,3,4,1,2,3,4,1,2,3,4,8,9,10,6,8,9,10,6,8,9,10,6,1,2,3,4,4,2,3,4,8,9,10,8,9,10,6,6,7};
        break;
    case 8:
        // # Config 8 - 7 & end forward
        pose_seq = {{0,0},{32, 0},{44, -15},{44,-30},{44,-65},{44,-95},{44,-115},{28,-127},{4.5,-128},{34,-129},{44,-143},{44,-158},{44,-198},{44, -241},{27, -256},{12,-256},{0, -256},{-70,-256},{-84,-241},{-84,-226},{-84.5,-199},{-84,-226},{-84,-241},{-94,-256},{-135.5,-256},{-182,-256},{-197,-256},{-210,-246},{-212,-195},{-212,-158},{-212,-145},{-200,-130},{-142,-128},{-202,-128},{-212,-113},{-212,-98},{-212,-74},{-212,-78},{-212,-63},{-200,-49},{-135,-48},{-94,-48},{-84,-33},{-84,-18},{-84,-13},{-103,0},{-114,0},{-135,1}};
        goal_types = {0,8,9,10,6,6,9,10,6,1,2,3,4,1,2,3,4,1,2,3,4,8,9,10,6,8,9,10,6,8,9,10,6,1,2,3,4,4,2,3,4,1,2,3,1,2,3,7};
        break;
    case 9:
        pose_seq = {{0,0},{24,-0.5},{32, -0.5},{44, -20},{44,-35},{44,-65},{44,-95},{44,-113},{25,-127},{0,-128},{15,-128},{33,-129},{44,-143},{44,-158},{44,-198},{44,-230},{44, -241},{27, -256},{12,-256},{0, -256},{-59,-256},{-73,-256},{-84,-241},{-84,-226},{-84.5,-198},{-85,-235},{-84,-244},{-94,-255},{-135.5,-255},{-178,-253},{-196.5,-242.5},{-207,-218},{-211,-198},{-212,-158},{-212,-145},{-200,-130},{-135.5,-128},{-150,-128},{-190,-128},{-203,-128},{-212,-113},{-212,-98},{-212,-74},{-212,-78},{-212,-63},{-199,-49},{-135,-48},{-110,-48},{-93,-48},{-85,-33},{-85,-18},{-85,-13},{-103,0},{-114,0},{-135,1}};
        goal_types = {0,6,8,9,10,6,6,9,10,11,1,0,2,3,4,0,1,2,3,4,0,1,2,3,5,8,9,10,6,8,9,10,6,8,9,10,11,1,0,1,2,3,4,4,2,3,4,0,1,2,3,1,2,3,7};
        break;
    case 10:
        pose_seq = {{0,0},{24,-0.5},{32, -0.5},{44, -20},{44,-35},{44,-65},{44,-95},{44,-113},{25,-127},{0,-128},{15,-128},{33,-129},{44,-143},{44,-158},{44,-198},{44,-230},{44, -241},{27, -256},{12,-256},{0, -256},{-59,-256},{-73,-256},{-84,-241},{-84,-226},{-84.5,-198},{-85,-235},{-91.8,-247.8},{-105,-255.5},{-135.5,-255},{-178,-253},{-196.5,-242.5},{-207,-218},{-211,-198},{-212,-158},{-212,-145},{-200,-130},{-135.5,-128},{-150,-128},{-190,-128},{-203,-128},{-212,-113},{-212,-98},{-212,-74},{-212,-78},{-212,-63},{-199,-49},{-135,-48},{-110,-48},{-93,-48},{-85,-33},{-85,-18},{-85,-13},{-103,0},{-114,0},{-135,1}};
        goal_types = {0,6,8,9,10,6,6,9,10,11,1,0,2,3,4,0,1,2,3,4,0,1,2,3,5,8,9,10,6,8,9,10,6,8,9,10,11,1,0,1,2,3,4,4,2,3,4,0,1,2,3,1,2,3,7};
        break;
  }
  geometry_msgs::Vector3 goals;
  goals.x = pose_seq[0][0];
  goals.y = pose_seq[0][1];
  goals.z = goal_types[0];
  ROS_INFO("Sending goals achievements");
  cont_pub.publish(goals);
}

void Goal::odom(const nav_msgs::Odometry::ConstPtr &msg) {
  int rad = 5;

  if (goal_cnt < pose_seq.size()) {
    if (goal_types[goal_cnt] == 5 || goal_types[goal_cnt] == 11)
    {
      rad = 3;
    }
    geometry_msgs::Vector3 goals;
    double diff = sqrt(pow(msg->pose.pose.position.x - pose_seq[goal_cnt][0],2) + pow(msg->pose.pose.position.y - pose_seq[goal_cnt][1],2) + pow(msg->pose.pose.position.z,2));
    if (diff < rad) {
      goal_cnt += 1;
      if (goal_cnt < pose_seq.size()) {
        goals.x = pose_seq[goal_cnt][0];
        goals.y = pose_seq[goal_cnt][1];
        goals.z = goal_types[goal_cnt];
        cont_pub.publish(goals);
      }
    }
  }
  else {
    return;
  }
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "goal_post");
  ros::NodeHandle n;
  Goal goal = Goal(&n);
  goal.callback();
  ros::spin();
  // ros::MultiThreadedSpinner spinner(3);
  // spinner.spin();
  return 0;
}

File: catkin_ws/airsim_2dnav/scripts/goals.py
#!/usr/bin/env python
# license removed for brevity
import rospy
import math
import tf
import actionlib
from std_msgs.msg import Header, Int8
from std_srvs.srv import Empty
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal
from actionlib_msgs.msg import GoalStatus
from sensor_msgs.msg import PointCloud
from geometry_msgs.msg import Pose, Point, Quaternion, PoseStamped, Vector3
from nav_msgs.msg import Odometry, OccupancyGrid
from tf.transformations import quaternion_from_euler
import roslaunch
import rospkg


class Goal:

    def __init__(self):
        self.pose_seq = list()
        self.goal_cnt = 0
        self.client = actionlib.SimpleActionClient('move_base', MoveBaseAction)
    def callback(self):
        pub = rospy.Publisher("/goal_type",Int8,queue_size=10)
        goal_type = Int8()
        rospy.loginfo("Waiting for move_base action server...")
        wait = self.client.wait_for_server(rospy.Duration(1000.0))
        if not wait:
            rospy.logerr("Action server not available!")
            rospy.signal_shutdown("Action server not available!")
            return
        rospy.loginfo("Connected to move base server")
        mb_goal = MoveBaseGoal()
        rospy.sleep(3) # Delay start
        rospy.loginfo("Starting goals achievements ...")

        # Start to 11th (last) goal - note some goals (4th and 8th) are not located where they're supposed to be as this allows the car to remain fast while having enough time to brake without adding more types of goals
        # U-turn
        points = [[-80, 0], [-135, 0], [-202, 0], [-212, -15], [-212, -30], [-212, -74], [-212,-118],[-197,-128],[-179,-128.5],[-135.5,-128],[-94,-128],[-84,-143],[-84,-153],[-84,-175],[-83,-195],[-79,-203],[-78, -207],[-84, -208], [-85, -205], [-84,-195],[-84,-128],[-84,-68],[-84,-58],[-99,-48],[-115,-48],[-135,-48],[-202, -48],[-212, -63],[-212, -78],[-212,-198],[-212, -241],[-212, -246],[-197, -256],[-182, -256],[-135.5, -256],[-20, -256],[34.5, -256],[44, -241],[44, -220],[44, -198],[44, -138],[29, -128],[1,-128],[7.5, -133],[15, -128],[34, -128],[44, -113], [44,-98], [44, -66]]

        orient_seq = [0, 0, 0,0,0, 0,0,0,0,0,0,0,0]
        goal_type.data = 0

        # List of goal quaternions:
        quat_seq = list()
        for yawangle in orient_seq:
            quat_seq.append(Quaternion(*(quaternion_from_euler(0, 0, yawangle * math.pi / 180, axes='sxyz'))))
        for i in range(len(points)):
            self.pose_seq.append(Pose(Point(points[i][0], points[i][1], 0), quat_seq[1]))
        mb_goal.target_pose.header = Header()
        mb_goal.target_pose.header.stamp = rospy.Time.now()
        mb_goal.target_pose.header.frame_id = "world_enu"
        mb_goal.target_pose.pose = self.pose_seq[self.goal_cnt]
        rospy.loginfo("Sending goal pose " + str(self.goal_cnt + 1) + " to Action Server")
        pub.publish(goal_type)
        self.client.send_goal(mb_goal)
    def position(self, msg):
        # Array of goal types
        # goal_arr = [0,4,1,2,3,4,1,2,3,4,1,2,3,4,1,5,5,5,5,3,0,1,2,3,4,1,2,3,4,1,2,3,4,4,1,2,3,4,1,2,7,6,5,1,2,3,7] # 1-11
        # # & Fast
        goal_arr = [0,4,1,2,3,4,1,2,3,4,1,2,3,4,1,5,5,5,5,3,0,1,1,2,3,4,1,2,3,4,1,1,2,3,4,4,1,2,3,4,1,2,7,6,5,1,2,3,7] # 1-11

        if goal_arr[self.goal_cnt] in [5,6,7]: # Smaller radius for 3-point D
            rad = 3
        else:
            rad = 5
        pub = rospy.Publisher("/goal_type",Int8,queue_size=10)
        goal_type = Int8()
        self.x = msg.pose.pose.position.x
        self.y = msg.pose.pose.position.y
        self.orient = math.sinh(msg.pose.pose.orientation.z)*2
        diff = math.sqrt((self.x - self.pose_seq[self.goal_cnt].position.x)**2 + (self.y - self.pose_seq[self.goal_cnt].position.y)**2)
        if diff < rad:
            self.goal_cnt += 1
            if self.goal_cnt < len(self.pose_seq):
                goal_type.data = goal_arr[self.goal_cnt]
                next_goal = MoveBaseGoal()
                next_goal.target_pose.header.frame_id = "world_enu"
                next_goal.target_pose.header.stamp = rospy.Time.now()
                next_goal.target_pose.pose = self.pose_seq[self.goal_cnt]
                rospy.loginfo("Sending goal pose " + str(self.goal_cnt + 1) + " to Action Server")
                rospy.loginfo("Now heading to goal %d", self.goal_cnt + 1)
                pub.publish(goal_type)
                self.client.send_goal(next_goal)
            else:
                launch.shutdown()
                rospy.loginfo("Final goal pose reached!")
                rospy.signal_shutdown("Final goal pose reached!")
                return
def listener():
    global msg, costmap, launch
    msg = Odometry
    goal = Goal()
    costmap = OccupancyGrid
    rospack = rospkg.RosPack()
    nav_dir=rospack.get_path('airsim_2dnav')
    uuid = roslaunch.rlutil.get_or_generate_uuid(None, False)
    roslaunch.configure_logging(uuid)
    launch = roslaunch.parent.ROSLaunchParent(uuid, [nav_dir+"/launch/move_base.launch"])
    launch.start()
    rospy.init_node("goal_post")
    goal.callback()
    rospy.Subscriber("/odom", Odometry, goal.position)
    rospy.spin()

if __name__ == '__main__':
    listener()
